pkg/ccl/changefeedccl/encoder_protobuf.go

```
@@ -0,0 +1,276 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package changefeedccl
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdcevent"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedpb"
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+	"github.com/gogo/protobuf/types"
+)
+
+type protobufEncoder struct {
+	updatedField       bool
+	mvccTimestampField bool
+	beforeField        bool
+	keyInValue         bool
+	topicInValue       bool
+	envelopeType       changefeedbase.EnvelopeType
+	targets            changefeedbase.Targets
+}
+
+// protobufEncoderOptions wraps EncodingOptions for initializing a protobufEncoder.
+type protobufEncoderOptions struct {
+	changefeedbase.EncodingOptions
+}
+
+var _ Encoder = &protobufEncoder{}
+
+// newProtobufEncoder constructs a new protobufEncoder from the given options and targets.
+func newProtobufEncoder(
+	ctx context.Context, opts protobufEncoderOptions, targets changefeedbase.Targets,
+) (Encoder, error) {
+	return &protobufEncoder{
+		envelopeType:       opts.Envelope,
+		keyInValue:         opts.KeyInValue,
+		topicInValue:       opts.TopicInValue,
+		beforeField:        opts.Diff,
+		updatedField:       opts.UpdatedTimestamps,
+		mvccTimestampField: opts.MVCCTimestamps,
+		targets:            targets,
+	}, nil
```

nit: dont need to return an error here